pkg/await/extensions_ingress.go

```diff
@@ -0,0 +1,362 @@
+package await
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi-kubernetes/pkg/client"
+	"github.com/pulumi/pulumi/pkg/diag"
+	"k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+)
+
+// ------------------------------------------------------------------------------------------------
+
+// Await logic for extensions/v1beta1/Ingress.
+//
+// The goal of this code is to provide a fine-grained account of the status of a Kubernetes Ingress
+// resource as it is being initialized. The idea is that if something goes wrong early, we want
+// to alert the user so they can cancel the operation instead of waiting for timeout (~10 minutes).
+//
+// The design of this awaiter is fundamentally an event loop on four channels:
+//
+//   1. The Ingress channel, to which the Kubernetes API server will proactively push every change
+//      (additions, modifications, deletions) to any Ingress it knows about.
+//   2. The Endpoint channel, which is the same idea as the Ingress channel, except it gets updates
+//      to Endpoint objects.
+//   3. A timeout channel, which fires after some time.
+//   4. A cancellation channel, with which the user can signal cancellation (e.g., using SIGINT).
+//
+// The `ingressInitAwaiter` will synchronously process events from the union of all these channels.
+// Any time the success conditions described above a reached, we will terminate the awaiter.
+//
+// x-refs:
+//   * https://github.com/nginxinc/kubernetes-ingress/blob/5847d1f3906287d2771f3767d61c15ac02522caa/docs/report-ingress-status.md
+
+// ------------------------------------------------------------------------------------------------
+
+type ingressInitAwaiter struct {
+	config           createAwaitConfig
+	ingress          *unstructured.Unstructured
+	ingressReady     bool
+	ingressClient    *dynamic.ResourceInterface
```

Comment:
We're doing this because of `initializeClient` right? I tend to find this awkward. Because we are using the client pool under the covers, it doesn't seem that bad to always just create the client any time that method is called.